List View: Fix performance issue when selecting all blocks

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of: https://github.com/WordPress/gutenberg/issues/49563

Fix a performance issue with the List View when selecting all blocks within the editor canvas, with the list view open. For long posts or pages, the current performance issue means it will likely feel broken as it can take many seconds to apply the selection.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

<!--
Currently, when a user goes to select all blocks from the editor canvas (press CMD+A a couple of times from within the editor canvas), if the list view is open, it can be very slow to update on a long post or page. This is because if the post is long enough to invoke the windowing logic for the list view, then currently changing the block selection will cause the non-displayed list view items to all be swapped out for real list view items, resulting in many hundreds of components being mounted.

The existing rule to render "real" list view items for selected blocks even if they're outside the list view window was introduced back in https://github.com/WordPress/gutenberg/pull/46895. The reason it was introduced was so that if a user selects a single block that is outside the current window (e.g. if the list view is scrolled to the top and a user selects a nested block at the very bottom of the post) then the "scroll into view" behaviour still needs to fire. This can only occur if the real list view item is used.

However, we don't need _all_ selected blocks to be rendered in order for that to work, just the one will do. Switching over to only rendering a selected block if it's outside the window if it's the only selected block, appears to do the trick. -->

There are two related performance issues on `trunk` that affect this behaviour:

* Currently all selected blocks are rendered as "real" list view items, bypassing the windowing logic, which means that there are unnecessary renders / mounts of selected blocks.
* The block item (`ListViewBlock`) is currently responsible for directing focus to it when the list view is mounted. Because it occurs at the block item level, it is possible for dozens of calls to competing `cellRef.current.focus()` to occur at once, causing the editor to slow down until they are all complete.

The result of these two issues interacting is that when you select all blocks with the list view open, (or go to open the list view after selecting all blocks), the list view slows to a crawl until all the `.focus()` calls have completed.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

1. Only render list view items for selected blocks that exist outside the current windowing logic if it is the first of the block selection. This prevents a large amount of re-rendering / re-mounting when a user goes to select all blocks.
2. Move the focus logic up to the list view root, so that it is only fired once, irrespective of the number of list view items rendered. To do so, I've moved the existing logic for focusing a list view item to a shared function in `utils` (the logic was introduced back in https://github.com/WordPress/gutenberg/issues/50422).

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Add a post or page with heaps of blocks (e.g. for testing that makes all this clear, add > 150 paragraphs)
3. Ensure the list view is open
4. Focus somewhere within a block in the editor canvas
5. Press CMD+A once to select all of that block. Press CMD+A again to select all blocks (if you're in a nested block, it'll just select the parent — keep pressing to eventually select all blocks, one level of hierarchy at a time)
6. If you're on `trunk`, it'll take a _long_ time to select all blocks in a long post or page
7. With this PR applied, it should take under a second to select all blocks. Maybe not snappy, but fast enough that it doesn't feel broken
8. Ensure scrolling up and down the list view remains performant (there'll still be a slight delay rendering items, caused by the windowing logic, but there shouldn't be any big slowdowns)

Assuming all this is working okay for this PR, we then need to make sure that we haven't broken the scroll into view logic when single block selection changes (introduced in https://github.com/WordPress/gutenberg/pull/46895). Follow the testing instructions from that PR:

> In a really long post or page, with the list view opened, scroll the editor canvas to far down the page or post and select a block outside of the currently visible area within the list view. The list view should scroll the selected block into view.
> To test the windowed area versus the non-windowed area, add a very large number of blocks (e.g. 60+) and try selecting from the very top block to the very bottom block in the list, via the editor canvas.

Finally, double check that focus still appears to be correct when deleting blocks via the list view (with an item in the list view focused, press the Backspace key).

## Screenshots or screencast <!-- if applicable -->

Note: the List View must be open to test out this performance issue.

The following screengrabs are from a post containing 203 blocks. The first press of CMD+A works quickly to select the individual block. It's the second press of CMD+A to select all blocks that's the problem here!

### Before (takes 8 seconds or so to select all)

https://github.com/WordPress/gutenberg/assets/14988353/700f8d73-4194-4f44-b635-f5b08604b3c4

### After (takes < 1 second to select all)

https://github.com/WordPress/gutenberg/assets/14988353/9cf84750-b9f4-45e4-9267-ee9e3ec562ac

andrewserong committed:
List View: Fix performance issue when selecting all blocks within the editor canvas in long posts

andrewserong committed:
Add a comment, rename const

apeatling reviewed (commented):
The select with 170 blocks worked in Chrome in less than a second, however in Safari the selection of all blocks did not work at all. No errors in the console though. 🤔

apeatling commented:
As an aside not for this PR, it is not easy to select all blocks using only the list view. I would expect to be able to select a block in the list view and use Cmd+A to select all the blocks in the list view. I also expected a "Select All" in the dropdown menu.

ramonjd commented:
Selecting is at least 5x faster than trunk! 👍🏻 

Here's what I'm seeing when scrolling after selection:


https://github.com/WordPress/gutenberg/assets/6458278/ac8b53b5-befa-4919-8971-d235721d5ae4

So while I have the impression I can select multiple blocks more quickly, if I want to browse the selected blocks, the lag seems to be deferred.

I have no idea if this is possible, but is there a way to batch selection? So, show block that pass `blockInView` and then then process the rest?

Mamaduka commented:
@apeatling, there's a separate PR for "Select all" in the list view. See #54899.

andrewserong commented:
Thanks for testing, everyone!

> @apeatling, there's a separate PR for "Select all" in the list view. See https://github.com/WordPress/gutenberg/pull/54899.

Yes, I split out this PR to see if we can land a performance improvement for 6.4, leaving the feature enhancement for 6.5. Assuming we can get this PR working adequately, of course 🙂

> So while I have the impression I can select multiple blocks more quickly, if I want to browse the selected blocks, the lag seems to be deferred.

Oh, great catch, I didn't notice the windowing slowing down so much like that. Given that selecting all blocks is a little hidden as an action, I think it'd be worth digging into the performance a little more here rather than unintentionally expose a performance issue that could be more prominent than what's currently on trunk 🤔

> I have no idea if this is possible, but is there a way to batch selection? So, show block that pass blockInView and then then process the rest?

Great question, I have no idea either! 😄
It's an interesting idea, though, I'll do a little more digging into the options.

Thanks folks, all that feedback gives me more avenues to explore!

andrewserong commented:
Oh, this has been a fun one to investigate! Thanks @ramonjd, your observation about the slowdown when scrolling the list view highlighted the real culprit, which is that `cellRef.current.focus()` is called _way_ too frequently [here](https://github.com/WordPress/gutenberg/blob/7a4330dd6d7fcc1ba337649a597f43877865f882/packages/block-editor/src/components/list-view/block.js#L150) (once for every newly mounted selected block), which appears to be slowing everything down. The `isTreeGridMounted` value appears to be stale, so we need to move things around a little bit to get things working properly again while maintaining performance.

I ran out of time to put up a good fix today, but I'll continue digging in tomorrow... I'll be keen to see if we can hopefully tidy up the focusing behaviour. Overall, though, if we use the fix in this PR (don't render selected blocks we don't need to), coupled with avoiding calling `.focus()` when not needed, I think we'll hopefully have something workable.

andrewserong committed:
Move block focus to be performed only once at the root of the list view, instead of within each block

andrewserong commented:
Update: I think I've come up with a reasonable fix in https://github.com/WordPress/gutenberg/pull/54900/commits/d4f40db7494519281d74efd9ca2cbd5ffbd4cf7f. Broadly, it does the following:

* Move the `.focus()` logic up to happen at the list view root on mount, instead of at the individual list view item level (`ListViewBlock`). I think this makes it easier to reason about, and it avoids the main cause of the slowdowns we were experiencing, which is when many separate `ListViewBlock`s (i.e. dozens) attempted to call focus to themselves at the same time.
* To achieve the above while preserving scroll into view behaviour, I've adjusted the logic of when to render the selected block slightly — we now render the selected block even if it isn't within the window, if it's the first block of the selection.
* To perform the focus on mount behaviour, I've borrowed the focus logic that was introduced in https://github.com/WordPress/gutenberg/pull/50422 and moved it to a shared util.

All up, I think this improves the performance with selected blocks overall, and in manual testing it appears to be better in the following cases:

* Select all blocks with the list view opened
* With the list view closed, select all blocks, and then open the list view (opening the list view is now faster)
* Windowing logic now correctly applies to selected blocks when scrolling up and down the list view

This will likely need some more manual testing (especially across browsers) to make sure it hasn't introduced any regressions, but all up, I _think_ we've honed in on a good performance improvement now.

As always, happy for any feedback, especially if there are concerns with the change! I'll ping a couple more folks as reviewers, just for visibility 🙂

apeatling reviewed (approved):
Re-tested and this is working great on Safari and Chrome, both very fast with 170 blocks. I didn't experience any scrolling issues after the latest fixes.

ramonjd commented:
Great find @andrewserong !

I've only tested in Chrome so far, but just from profiling I can tell that the focus() isn't being constantly called when rendering the hidden items:

### Before

There's a constant and solid call to focus

<img width="652" alt="Screenshot 2023-10-04 at 11 58 55 am" src="https://github.com/WordPress/gutenberg/assets/6458278/c2ad2a28-83dc-4afc-a6ac-db53cc2044b4">


### After

The "dip" is the point where the hidden items are being rendered on scroll.

<img width="567" alt="Screenshot 2023-10-04 at 11 53 03 am" src="https://github.com/WordPress/gutenberg/assets/6458278/d498397e-6a69-4b3f-af2a-9a294e358ce8">

Screencast

![2023-10-04 11 53 20](https://github.com/WordPress/gutenberg/assets/6458278/390bfe4d-e46f-4bc8-bee5-49ccd6bead59)

kevin940726 reviewed (approved):
Works well for me! I believe the bug was introduced by me 🤦. Thank you! 💯

andrewserong commented:
Thanks for the reviews, everyone! 🙇 

> Works well for me! I believe the bug was introduced by me 🤦. Thank you! 💯

No, no, I don't think it was you! I think we both merged PRs that might have exposed the bug a little bit, but it turns out the underlying issue dates right back to the introduction of the persistent list view in https://github.com/WordPress/gutenberg/pull/28637 as the logic there (at the block level, instead of at the root of the component) could result in many simultaneous (or close to it) calls to `.focus()`, and the subsequent windowing logic has been tricky to wrestle with when combined with block-level rules. So it's been a bit tangled for a long while 😄

Hopefully this neatens things up a little bit! I'll merge it in now, happy to look into any follow-ups if we run into any further issues.

mikachan commented:
I just cherry-picked this PR to the 6.4-beta3 branch to get it included in the next release: fba6504a45
